proposals/NNNN-async-function-isolation.md

```diff
@@ -0,0 +1,897 @@
+nonisolated async function that will switch off the actor to run, which would
+lead to a data race if actor-isolated state is passed to the `body` parameter.
+
+Moreover, the above explanation of isolation rules for async closures is
+extremely difficult to understand; the default isolation rules are too
+complicated.
+
+## Proposed solution
+
+I propose changing nonisolated async functions to inherit the isolation of the
+caller by default. This means that nonisolated functions always have the same
+isolation rules, regardless of whether the function is synchronous or
+asynchronous. This makes the following example from the motivation section
+valid, because the call to `x.performAsync()` does not cross an isolation
+boundary:
+
+```swift
+class NotSendable {
+  func performSync() { ... }
+  func performAsync() async { ... }
+}
+
+actor MyActor {
+  let x: NotSendable
+
+  func call() async {
+    x.performSync() // okay
+
+    await x.performAsync() // okay
+  }
+}
+```
+
+This proposal also introduces the `@concurrent` declaration attribute to opt
+out of isolation inheritance, so that the function always switches off of an
+actor to run.
+
+## Detailed design
+
+### Nonisolated async functions
+
+Nonisolated async functions inherit the isolation of the caller:
+
+```swift
+class NotSendable {
+  func performSync() { ... }
+  func performAsync() async { ... }
+}
+
+actor MyActor {
+  let x: NotSendable
+
+  func call() async {
+    x.performSync() // okay
+
+    await x.performAsync() // okay
+  }
+}
+```
+
+In the above code, the call to `x.performAsync()` continues running on the
+`self` actor instance. The code does not produce a data-race safety error,
+because the `NotSendable` instance `x` does not leave the actor.
+
+This behavior is accomplished by implicitly passing an optional actor parameter
+to the async function. The function will run on this actor's executor. See the
+[Executor switching](#executor-switching) section for more details on why the
+actor parameter is necessary.
+
+The implicit parameter is not preserved when using a nonisolated async function
+as a value. When referencing a nonisolated async function unapplied in a
+context that expects a nonisolated `@Sendable` or `sending` function type, the
+function will switch off of the caller's actor when the function value is
+called, and sendable checking will be applied to argument and result values.
+
+> Note: It is not feasible to implicitly add parameters to function values
+> without widespread ABI impact. It's possible to stage in an ABI change for
+> a function declaration itself; see the
+> [ABI Compatibility](#abi-compatibility) section for more information.
+
+For example:
+
+```swift
+class NotSendable { ... }
+
+func useAsValue(_ ns: NotSendable) async { ... }
+
+@MainActor let global: NotSendable = .init()
+
+@MainActor
+func callSendableClosure(closure: @Sendable (NotSendable) async -> Void) {
```

Comment:
I *think* this function signature should be changed to:

`func callSendableClosure(_ closure: @Sendable (NotSendable) async -> Void) async`

Reply:
Indeed, fixed!